### PR TITLE
Specific interface for provider enhancement (2/3)

### PR DESCRIPTION
**Pull request by colerogers**
Change for the the specific billing interface. Exposes `onPlanUpdatePublished` & `onAutomatedPlanUpdatePublished` from the `v2/alerts/billing` namespace

Ex ~ 
```
import {
  onPlanUpdatePublished,
  onAutomatedPlanUpdatePublished
} from "firebase-functions/v2/alerts/billing";

// planUpdatePublished interface
export const func1 = onPlanUpdatePublished((event) => { … });
export const func2 = onPlanUpdatePublished({ region: "us-west1" }, (event) => { … });
// robotPlanUpdatePublished interface
export const func3 = onAutomatedPlanUpdatePublished((event) => { … });
export const func5 = onAutomatedPlanUpdatePublished({ region: "us-west1" }, (event) => { … });
```

### CODE REVIEW

File: src/v2/providers/alerts/billing.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+/** @internal */
+export function onOperation<T>(
+  alertType: string,
+  optsOrHandler:
+    | options.EventHandlerOptions
+    | ((event: BillingEvent<T>) => any | Promise<any>),
+  handler: (event: BillingEvent<T>) => any | Promise<any>
+): CloudFunction<FirebaseAlertData<T>> {
+  if (typeof optsOrHandler === 'function') {
+    handler = optsOrHandler as (event: BillingEvent<T>) => any | Promise<any>;
```
Comment (taeold) [inline src/v2/providers/alerts/billing.ts:88]: Woo `(event: BillingEvent<T>) => any | Promise<any>;` is used thrice! This meets the threshold for DRY for me:

```ts
type BillingEventHandler<T> = (event: BillingEvent<T>) => any | Promise<any>;
```

File: src/v2/providers/alerts/billing.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    return handler(raw as BillingEvent<T>);
+  };
+
+  func.run = handler;
+
+  // TypeScript doesn't recognize defineProperty as adding a property and complains
+  // that __endpoint doesn't exist. We can either cast to any and lose all type safety
+  // or we can just assign a meaningless value before calling defineProperty.
+  func.__trigger = 'silence the transpiler';
+  func.__endpoint = {} as ManifestEndpoint;
```
Comment (taeold) [inline src/v2/providers/alerts/billing.ts:102]: Can be removed.

File: src/v2/providers/alerts/billing.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  defineEndpoint(func, optsOrHandler, alertType, undefined);
```
Comment (taeold) [inline src/v2/providers/alerts/billing.ts:103]:
```suggestion
  defineEndpoint(func, optsOrHandler, alertType);
```

File: src/v2/providers/alerts/billing.ts (hunk 1 above)
Comment (colerogers) [reply src/v2/providers/alerts/billing.ts:95]: Oh that looks way better, thanks for the suggestion!

File: src/v2/providers/alerts/billing.ts (hunk 2 above)
Comment (colerogers) [reply src/v2/providers/alerts/billing.ts]: removed

File: src/v2/providers/alerts/billing.ts (hunk 3 above)
Comment (colerogers) [reply src/v2/providers/alerts/billing.ts]: removed

**Review by taeold — APPROVED** [review]
🔥 Great stuff.

File: src/v2/providers/alerts/billing.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+ */
+export type BillingEvent<T> = CloudEvent<FirebaseAlertData<T>, WithAlertType>;
+
+/** @internal */
+export const planUpdateAlert = 'billing.planUpdate';
+/** @internal */
+export const automatedPlanUpdateAlert = 'billing.automatedPlanUpdate';
+
+/** @internal */
+type BillingEventHandler<T> = (event: BillingEvent<T>) => any | Promise<any>;
```
Comment (inlined) [inline src/v2/providers/alerts/billing.ts:38]: This typedef makes code complete worse (it asks for a BillingEventHandler instead of a function signature)

**Review by inlined — APPROVED** [review]

File: src/v2/providers/alerts/billing.ts (hunk 4 above)
Comment (colerogers) [reply src/v2/providers/alerts/billing.ts:38]: Oh shoot, we should probably change it back for better usability then, right?

Comment (taeold) [reply src/v2/providers/alerts/billing.ts:38]: I wanted to test out the automcomplete and see what shows, but I couldn't get to any of the billing handlers at all.

@colerogers Can you double check if this is working for you?

1) `npm run build:pack`
2) Install the generated tgz locally on your function
3) Try to get import to work.

Comment (taeold) [reply src/v2/providers/alerts/billing.ts:38]: In fact, I couldn't even get `lib/v2/providers/billing.js` to show up. what gives?

Comment (colerogers) [reply src/v2/providers/alerts/billing.ts:38]: @taeold I was able to get it to work, just pushed an update. Seems like we can't rely on the package.json export path alone, we need to also export from the root index.ts. I think there was a comment about exporting on one of the other PRs and I ended up changing it on all of them

Comment (colerogers) [reply src/v2/providers/alerts/billing.ts]: @inlined Removed BillingEventHandler type from the latest commit

File: src/v2/providers/alerts/billing.ts (hunk 1 above)
Comment (colerogers) [reply src/v2/providers/alerts/billing.ts:98]: @taeold Removed based on typing feedback below

**Pull request merged**